File: Java/src/main/java/aoc/day1/Day01.java
package aoc.day1;

import aoc.template.DaySolution;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.PriorityQueue;

public class Day01 extends DaySolution {
    // Part 1 fields
    private PriorityQueue<Integer> leftSide;
    private PriorityQueue<Integer> rightSide;

    // Part 2 fields
    private List<Integer> leftSideNumbers;
    private Map<Integer, Integer> numberToOccurrence;

    public Day01() {
        super(1);
    }

    @Override
    public String solvePart1(List<String> input) {
        // Initialize data structures for part 1
        initializePartOneStructures();

        // Parse input and populate data structures
        parseInputForPartOne(input);

        // Calculate total distance
        long totalDistance = calculateTotalDistance();

        return Long.toString(totalDistance);
    }

    private void initializePartOneStructures() {
        leftSide = new PriorityQueue<>();
        rightSide = new PriorityQueue<>();
    }

    private void parseInputForPartOne(List<String> input) {
        for (String line : input) {
            String[] nums = line.split("\\s+");

            if (nums.length >= 2) {
                leftSide.add(Integer.parseInt(nums[0]));
                rightSide.add(Integer.parseInt(nums[1]));
            }
        }
    }

    private long calculateTotalDistance() {
        long totalDistance = 0;

        while (!leftSide.isEmpty() && !rightSide.isEmpty()) {
            int leftValue = leftSide.poll();
            int rightValue = rightSide.poll();
            int distance = Math.abs(rightValue - leftValue);

            totalDistance += distance;
        }

        return totalDistance;
    }

    @Override
    public String solvePart2(List<String> input) {
        // Initialize data structures for part 2
        initializePartTwoStructures();

        // Parse input and populate data structures
        parseInputForPartTwo(input);

        // Calculate the result based on occurrences
        long result = calculateOccurrenceBasedResult();

        return Long.toString(result);
    }

    private void initializePartTwoStructures() {
        leftSideNumbers = new ArrayList<>();
        numberToOccurrence = new HashMap<>();
    }

    private void parseInputForPartTwo(List<String> input) {
        for (String line : input) {
            String[] nums = line.split("\\s+");

            if (nums.length >= 2) {
                int leftValue = Integer.parseInt(nums[0]);
                int rightValue = Integer.parseInt(nums[1]);

                leftSideNumbers.add(leftValue);
                numberToOccurrence.compute(rightValue, (key, count) -> count == null ? 1 : count + 1);
            }
        }
    }

    private long calculateOccurrenceBasedResult() {
        long result = 0;
        for (Integer leftValue : leftSideNumbers) {
            Integer occurrences = numberToOccurrence.getOrDefault(leftValue, 0);
            result += (long) leftValue * occurrences;
        }
        return result;
    }
}

File: C++/src/aoc/Day01.cpp
#include "Day01.hpp"
#include <iostream>
#include <sstream>
#include <algorithm>
#include <numeric>

namespace aoc {

Day01::Day01():
DaySolution(1)
{
}

std::string Day01::solvePart1(const std::vector<std::string>& input)
{
    // Initialize data structures for part 1
    initializePartOneStructures();

    // Parse input and populate data structures
    parseInputForPartOne(input);

    // Calculate total distance
    long totalDistance = calculateTotalDistance();

    std::ostringstream result;
    result << totalDistance;
    return result.str();
}

void Day01::initializePartOneStructures()
{
    // Clear any existing data
    while (!leftSide.empty()) leftSide.pop();
    while (!rightSide.empty()) rightSide.pop();
}

void Day01::parseInputForPartOne(const std::vector<std::string>& input)
{
    for (const auto& line : input) {
        std::istringstream iss(line);
        int leftValue, rightValue;
        
        if (iss >> leftValue >> rightValue) {
            leftSide.push(leftValue);
            rightSide.push(rightValue);
        }
    }
}

long Day01::calculateTotalDistance()
{
    long totalDistance = 0;

    while (!leftSide.empty() && !rightSide.empty()) {
        int leftValue = leftSide.top();
        leftSide.pop();
        
        int rightValue = rightSide.top();
        rightSide.pop();
        
        int distance = std::abs(rightValue - leftValue);
        totalDistance += distance;
    }

    return totalDistance;
}

std::string Day01::solvePart2(const std::vector<std::string>& input)
{
    // Initialize data structures for part 2
    initializePartTwoStructures();

    // Parse input and populate data structures
    parseInputForPartTwo(input);

    // Calculate the result based on occurrences
    long result = calculateOccurrenceBasedResult();

    std::ostringstream resultStream;
    resultStream << result;
    return resultStream.str();
}

void Day01::initializePartTwoStructures() 
{
    leftSideNumbers.clear();
    numberToOccurrence.clear();
}

void Day01::parseInputForPartTwo(const std::vector<std::string>& input)
{
    for (const auto& line : input) {
        std::istringstream iss(line);
        int leftValue, rightValue;
        
        if (iss >> leftValue >> rightValue) {
            leftSideNumbers.push_back(leftValue);
            numberToOccurrence[rightValue]++;
        }
    }
}

long Day01::calculateOccurrenceBasedResult() 
{
    long result = 0;
    for (const auto& leftValue : leftSideNumbers) {
        auto it = numberToOccurrence.find(leftValue);
        int occurrences = (it != numberToOccurrence.end()) ? it->second : 0;
        result += static_cast<long>(leftValue) * occurrences;
    }
    return result;
}

} // namespace aoc
